### PR TITLE
NXP-20626: Support RFC 3230/5843 Digest Headers in CMIS contentStream ops

### PR DESCRIPTION
**Pull request by ron1**

### CODE REVIEW

File: nuxeo-opencmis-impl/src/main/java/org/nuxeo/ecm/core/opencmis/impl/server/NuxeoContentStream.java [hunk 1]
```diff
@@ -114,6 +119,28 @@ public static boolean isHeadRequest(HttpServletRequest request) {
...
+        }
+        Enumeration<String> values = request.getHeaders("want-digest");
+        if (values == null) {
+            return false;
+        }
+        String digestAlgorithmUC = digestAlgorithm.toUpperCase();
+        Iterator it = Iterators.forEnumeration(values);
+        while (it.hasNext()) {
+            String value = ((String) it.next()).toUpperCase();
+            if (value.startsWith(digestAlgorithmUC)) {
```
Comment (efge) [inline nuxeo-opencmis-impl/src/main/java/org/nuxeo/ecm/core/opencmis/impl/server/NuxeoContentStream.java]: Why `startswith` and not straightforward `equalsIgnoreCase` comparison? If it's about the `;q=...` then it's better to split on `;` and use the first part for the actual digest.

File: nuxeo-opencmis-impl/src/main/java/org/nuxeo/ecm/core/opencmis/impl/server/NuxeoPropertyData.java [hunk 2]
```diff
@@ -299,7 +307,8 @@ public static ContentStream getContentStream(DocumentModel doc, HttpServletReque
         return NuxeoContentStream.create(doc, DownloadService.BLOBHOLDER_0, blob, "cmis", null, lastModified, request);
     }
 
-    public static void setContentStream(DocumentModel doc, ContentStream contentStream, boolean overwrite)
+    public static void setContentStream(DocumentModel doc, ContentStream contentStream, boolean overwrite,
+            HttpServletRequest request)
```
Comment (efge) [inline nuxeo-opencmis-impl/src/main/java/org/nuxeo/ecm/core/opencmis/impl/server/NuxeoPropertyData.java]: You added `request` but it is unused.

File: nuxeo-opencmis-impl/src/main/java/org/nuxeo/ecm/core/opencmis/impl/server/NuxeoPropertyData.java [hunk 3]
```diff
@@ -346,6 +355,73 @@ public static Blob getPersistentBlob(ContentStream contentStream, String filenam
...
+    }
+
+    protected static String extractDigestFromRequestHeaders(HttpServletRequest request, String digestAlgorithm) {
+        if (request == null) {
+            return null;
+        }
+        if (request instanceof HttpServletRequestWrapper) {
+            request = (HttpServletRequest) ((HttpServletRequestWrapper) request).getRequest();
+        }
+        Enumeration<String> digests = request.getHeaders("digest");
```
Comment (efge) [inline nuxeo-opencmis-impl/src/main/java/org/nuxeo/ecm/core/opencmis/impl/server/NuxeoPropertyData.java]: Use a constant for `"digest"`.

File: nuxeo-opencmis-tests/src/test/java/org/nuxeo/ecm/core/opencmis/impl/DummyCmisBlobProvider.java [hunk 4]
```diff
@@ -69,6 +71,25 @@ public String getProviderId() {
...
+            }
+
+            @Override
+            public String getDigest() {
+                String digest = super.getDigest();
+                if (digest == null) {
+                    try {
+                        return DigestUtils.md5Hex(getStream());
+                    } catch (IOException e) {
+                        return null;
```
Comment (efge) [inline nuxeo-opencmis-tests/src/test/java/org/nuxeo/ecm/core/opencmis/impl/DummyCmisBlobProvider.java]: If this shouldn't happen then just rethrow as a `RuntimeException`. If it's a legitimate case then please add a comment explaining when that happens.

File: nuxeo-opencmis-impl/src/main/java/org/nuxeo/ecm/core/opencmis/impl/server/NuxeoCmisService.java [hunk 5]
```diff
@@ -853,21 +863,46 @@ public AllowableActions getAllowableActions(String repositoryId, String objectId
...
+        return cs;
+    }
+
+    protected void setResponseHeader(String headerName, CallContext callContext, Blob blob) {
+        try {
+            String digest = NuxeoPropertyData.transcodeHexToBase64(blob.getDigest());
+            HttpServletResponse response = (HttpServletResponse) callContext.get(CallContext.HTTP_SERVLET_RESPONSE);
+            String digestPrefix = "Digest".equals(headerName) ? blob.getDigestAlgorithm() + "=" : "";
+            response.setHeader(headerName, digestPrefix + digest);
+        } catch (DecoderException ignored) {
```
Comment (efge) [inline nuxeo-opencmis-impl/src/main/java/org/nuxeo/ecm/core/opencmis/impl/server/NuxeoCmisService.java]: Don't swallow exceptions. Anyway, don't deal with `DecoderException` here, but in the `transcodeHexToBase64` when you can rethrow as a `RuntimeException` as it's a "shouldn't happen" case.

File: nuxeo-opencmis-impl/src/main/java/org/nuxeo/ecm/core/opencmis/impl/server/NuxeoContentStream.java [hunk 6]
```diff
@@ -114,6 +119,28 @@ public static boolean isHeadRequest(HttpServletRequest request) {
         return request.getMethod().equals("HEAD");
     }
 
+    public static boolean hasWantDigestRequestHeader(HttpServletRequest request, String digestAlgorithm) {
+        if (request == null || digestAlgorithm == null) {
+            return false;
+        }
+        if (request instanceof HttpServletRequestWrapper) {
+            request = (HttpServletRequest) ((HttpServletRequestWrapper) request).getRequest();
```
Comment (efge) [inline nuxeo-opencmis-impl/src/main/java/org/nuxeo/ecm/core/opencmis/impl/server/NuxeoContentStream.java]: Is the unwrap needed? Doesn't `request.getHeaders` still work? If not, then please add a comment explaining why.

File: nuxeo-opencmis-impl/src/main/java/org/nuxeo/ecm/core/opencmis/impl/server/NuxeoPropertyData.java [hunk 7]
```diff
@@ -346,6 +355,73 @@ public static Blob getPersistentBlob(ContentStream contentStream, String filenam
...
+                    blobDigest, reqDigest));
+        }
+    }
+
+    protected static String extractDigestFromRequestHeaders(HttpServletRequest request, String digestAlgorithm) {
+        if (request == null) {
+            return null;
+        }
+        if (request instanceof HttpServletRequestWrapper) {
+            request = (HttpServletRequest) ((HttpServletRequestWrapper) request).getRequest();
```
Comment (efge) [inline nuxeo-opencmis-impl/src/main/java/org/nuxeo/ecm/core/opencmis/impl/server/NuxeoPropertyData.java]: Same comment about unwrap.

File: nuxeo-opencmis-impl/src/main/java/org/nuxeo/ecm/core/opencmis/impl/server/NuxeoPropertyData.java [hunk 8]
```diff
@@ -346,6 +355,73 @@ public static Blob getPersistentBlob(ContentStream contentStream, String filenam
...
+        return null;
+    }
+
+    public static String transcodeBase64ToHex(String base64String){
+        byte[] bytes = Base64.decodeBase64(base64String);
+        String hexString = Hex.encodeHexString(bytes);
+        return hexString;
+    }
+
+    public static String transcodeHexToBase64(String hexString) throws DecoderException {
```
Comment (efge) [inline nuxeo-opencmis-impl/src/main/java/org/nuxeo/ecm/core/opencmis/impl/server/NuxeoPropertyData.java]: As mentioned before, deal with `DecoderException` inside the method, don't expose it to callers.

File: nuxeo-opencmis-impl/src/main/java/org/nuxeo/ecm/core/opencmis/impl/server/NuxeoCmisService.java [hunk 9]
```diff
@@ -623,6 +629,10 @@ protected NuxeoObjectData createObject(String repositoryId, Properties propertie
         } else {
             doc = coreSession.saveDocument(doc);
         }
+        if (setContentStream) {
+            Blob blob = coreSession.getDocument(doc.getRef()).getAdapter(BlobHolder.class).getBlob();
+            NuxeoPropertyData.validateBlobDigest(blob, request);
```
Comment (ron1) [inline nuxeo-opencmis-impl/src/main/java/org/nuxeo/ecm/core/opencmis/impl/server/NuxeoCmisService.java]: Great suggestion resulting in much cleaner code

**Comment by ron1** [comment]
PR comments have been addressed and some cleanup has been performed.

**Comment by efge** [comment]
We're currently testing before push f0276b00da838e2fee1fb57c51d8b4e6d29ef767 which is a cleaned up version of your code + a fix for `extractDigestFromRequestHeaders` which had a logic issue (ignored passed `digestAlgorithm`).

**Comment by efge** [comment]
Pushed as 67f409298302124e5d2f5b26a6176160990e698b

**Comment by efge** [comment]
FYI 88e60b9a9f9957f0e01c5a3b659d816775d53409 was also needed, your use of `Framework.trackFile` was incorrect.

**Pull request closed without merging**